lnrpc/routerrpc/router_backend.go

```diff
@@ -258,34 +256,136 @@ func (r *RouterBackend) parseQueryRoutesRequest(in *lnrpc.QueryRoutesRequest) (
 	if sourcePubKey != r.SelfNode {
 		maxTotalTimelock = math.MaxUint32
 	}
+
 	cltvLimit, err := ValidateCLTVLimit(in.CltvLimit, maxTotalTimelock)
 	if err != nil {
 		return nil, err
 	}
 
-	// We need to subtract the final delta before passing it into path
-	// finding. The optimal path is independent of the final cltv delta and
-	// the path finding algorithm is unaware of this value.
-	finalCLTVDelta := r.DefaultFinalCltvDelta
-	if in.FinalCltvDelta != 0 {
-		finalCLTVDelta = uint16(in.FinalCltvDelta)
-	}
+	// If we have a blinded path set, we'll get a few of our fields from
+	// inside of the path rather than the request's fields.
+	var (
+		targetPubKey   *route.Vertex
+		routeHintEdges map[route.Vertex][]*channeldb.CachedEdgePolicy
+		blindedPmt     *routing.BlindedPayment
+
+		// finalCLTVDelta varies depending on whether we're sending to
+		// a blinded route or an unblinded node. For blinded paths,
+		// our final cltv is already baked into the path so we restrict
+		// this value to zero on the API. Bolt11 invoices have a 
+		// default, so we'll fill that in for the non-blinded case.
+		finalCLTVDelta uint16
+
+		// destinationFeatures is the set of features for the
+		// destination node.
+		destinationFeatures *lnwire.FeatureVector
+	)
 
-	// Do bounds checking without block padding so we don't give routes
-	// that will leave the router in a zombie payment state.
-	err = routing.ValidateCLTVLimit(cltvLimit, finalCLTVDelta, false)
-	if err != nil {
-		return nil, err
-	}
+	// Validate that the fields provided in the request are sane depending
+	// on whether it is using a blinded path or not.
+	if len(in.BlindedPaymentPaths) > 0 { //nolint:nestif
+		if len(in.PubKey) != 0 {
+			return nil, fmt.Errorf("target pubkey: %x should "+
+				"not be set when blinded path is provided",
+				in.PubKey)
+		}
 
-	cltvLimit -= uint32(finalCLTVDelta)
+		if len(in.BlindedPaymentPaths) != 1 {
+			return nil, errors.New("query routes only supports " +
+				"a single blinded path")
+		}
 
-	// Parse destination feature bits.
-	features, err := UnmarshalFeatures(in.DestFeatures)
-	if err != nil {
-		return nil, err
+		var (
+			blindedPath = in.BlindedPaymentPaths[0]
+		)
+
+		if len(in.RouteHints) > 0 {
+			return nil, errors.New("route hints and blinded " +
+				"path can't both be set")
+		}
+
+		blindedPmt, err = unmarshalBlindedPayment(blindedPath)
+		if err != nil {
+			return nil, fmt.Errorf("unmarshal blinded payment: %w",
+				err)
+		}
+
+		if err := blindedPmt.Validate(); err != nil {
+			return nil, fmt.Errorf("invalid blinded path: %w", err)
+		}
+
+		if in.FinalCltvDelta != 0 {
+			return nil, errors.New("final cltv delta should be " +
+				"zero for blinded paths")
+		}
+
+		// Just apply block padding for blinded routes, as CLTV delta 
```

Comment:
This has to do with the block height that these expiries are expressed relative to - it's not actually a blinded paths issue, can happen with regular payments as well. Say we have: 
- Current Height = 100
- Final CLTV Expiry = 10 

### A) Same View of Chain
**Sender**: 
`cltv_expiry` for final incoming HTLC = `sender_current_height` + `final_cltv` + `sender_delta`
`cltv_expiry` for final incoming HTLC = 100 + 10 + 0 = 110

**Receiver**: 
Validate: `cltv_expiry >= receiver_current_height + min_final_cltv_expiry_delta` 
 Validate ✅: 110 >= 100 + 10  

### B) Different View of Chain (no delta)
Now, consider the case where the sender creates the route at height 100, sends the payment and by the time the recipient gets it we're at block 101.

**Sender**: 
`cltv_expiry` for final incoming HTLC = `sender_current_height` + `final_cltv` + `sender_delta`
`cltv_expiry` for final incoming HTLC = 100 + 10 + 0 = 110

**Receiver**: 
Validate: `cltv_expiry >= receiver_current_height + min_final_cltv_expiry_delta`
Validate ❌:  110 >= 101 + 10
-> validation will fail because it's comparing to a different current height!

For this reason, we want to set a `sender_delta` to allow for a block being mined (or the possibility that our view of the chain is a little behind) so that our payments still go through.  

### C) Different View of Chain (with delta)
For the same example above (B) where we send and receive with different block heights:

**Sender**: 
`cltv_expiry` for final incoming HTLC = `sender_current_height` + `final_cltv` + `sender_delta`
`cltv_expiry` for final incoming HLTC = 100 + 10 + 3 = 113

**Receiver**: 
Validate: `cltv_expiry >= receiver_current_height + min_final_cltv_expiry_delta` 
Validate ✅: 113 >= 101 + 10 

You can think of `final_cltv_delta` and the recipient saying "this is the number of blocks that I want from *my block height when I receive the HTLC* to be able to settle on chain", so we always need to make sure we give them _at least_ that delta.

Reply:
What's also pretty confusing here is that we don't add this `sender_delta` (/block padding) in the regular case of query routes (I'm not sure why - IMO it should have it, could be done separately).

Reply:
excellent explanation  🙏, I think it finally clicked. 

The main point is that the `min_final_cltv_expiry_delta ` and the `final_cltv` are different values for lnd.  

The `min_final_cltv_expiry_delta` for a lnd node is 13:
https://github.com/lightningnetwork/lnd/blob/e8c11b9898890857275fc25bd575fa0cace5db35/lncfg/config.go#L41

whereas the minimum `final_cltv_delta` we enforce is 18:

https://github.com/lightningnetwork/lnd/blob/e8c11b9898890857275fc25bd575fa0cace5db35/lnrpc/invoicesrpc/addinvoice.go#L366-L370

so the padding is baked in I guess. Maybe we also stick to this design for blinded paths as well. Meaning that we make sure when creating the blinded path that we do add a `blinded_path_cltv_delta` which enforces a `final_cltv_delta` of 18 and keep the `min_final_cltv_expiry_delta` at 13. 

One thing to keep in mind is that what happens to node_implementations when the `outgoing_cltv_value ` < `current_blockheight_sender`, normally for blinded paths this value should just adhere to the following requirement:

`cltv_expiry - cltv_expiry_delta >= outgoing_cltv_value` this will always be true if we include the `final_cltv_delta` in the `blinded_path_cltv_delta`.

But maybe some other implementations do check `outgoing_cltv_value` > `current_block_height` which is not part of the spec. So we could just add a 3 block padding to be absolutely sure. I am ok with both ways, as currently is or removing it and just let the creator of the blinded path assure it.

Reply:
> The main point is that the min_final_cltv_expiry_delta  and the final_cltv are different values for lnd.

Not sure I understand this statement? Both of these values are used for _receiving_, and this PR is only about sending..

> The min_final_cltv_expiry_delta for a lnd node is 13
> whereas the minimum final_cltv_delta we enforce is 18:

There's a difference between: 
- `min_final_cltv_expiry_delta`: the lowest CLTV delta we'll accept for an incoming HLTC (mostly spontaneous payments)
- `final_cltv_delta`: the CLTV delta we set for our invoices.

I think that the naming is tripping you up here. In the spec we [refer](https://github.com/lightning/bolts/blob/master/04-onion-routing.md#payload-for-the-last-node) to `min_final_cltv_delta` which is _actually_ `final_cltv` (ie, the value in the [invoice](https://github.com/lightning/bolts/blob/master/11-payment-encoding.md#tagged-fields)).

> so the padding is baked in I guess

The padding is *not* baked in. The HTLC that we receive must *exactly* match the information that we're provided in the onion. 

Eg: we create an invoice with final delta = 18, at height 100. Somebody goes to pay the invoice, creating a final HTLC that expires at height 118. We will not accept a HTLC for this invoice that expires at block 113, because the HTLC *must* match the delta that we gave (even though in other cases we'd accept something lower). Because the [spec](https://github.com/lightning/bolts/blob/master/04-onion-routing.md#requirements) says: 
`MUST return an error if incoming cltv_expiry < current_block_height + min_final_cltv_expiry_delta.`

> But maybe some other implementations do check outgoing_cltv_value > current_block_height which is not part of the spec. So we could just add a 3 block padding to be absolutely sure.

I think we should add the padding so that we have better payment success - if a block is mined during payment, we need to have the delta or the payment will fail?

Reply:
> The padding is not baked in. The HTLC that we receive must exactly match the information that we're provided in the onion.

Ohh you are right, did only think of `sendpayment` when analysing this. For the query route command we do not add the blockpadding when not using blinded paths. Maybe worth adding it in a separate PR but really minor issue ?


> I think we should add the padding so that we have better payment success - if a block is mined during payment, we need to have the delta or the payment will fail?

Yes I think your approach is the right one here. We should add a padding in blinded paths as well. One could the padding on the receiver add compared to the sender but I like the approach way more on the sender side.


So everything looks good now just the minor issue we discussed where we want the ability to add a cltv_delta in case the blinded path has only 1 hop?

Reply:
> So everything looks good now just the minor issue we discussed where we want the ability to add a cltv_delta in case the blinded path has only 1 hop?

Yeah think so! Adding that in shortly. 

Re block padding: going to stay consistent with how we have it now and *not* have block padding for blinded routes (discussed offline - it's confusing to have a difference here). Instead, going to follow this PR up with a new `block_padding` field for the `QueryRoutesRequest` so that senders can opt in.